### PR TITLE
Do not create a cache field for lambda if it depends on caller's type argument

### PR DESCRIPTION
**Pull request by nbalakin**
Fix for #44720

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 1]
```diff
@@ -1597,17 +1598,23 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
...
-                            topLevelMethodId.Generation,
-                            lambdaId.Ordinal,
-                            lambdaId.Generation);
-
-                        var cacheField = new SynthesizedLambdaCacheFieldSymbol(translatedLambdaContainer, cacheVariableType, cacheVariableName, _topLevelMethod, isReadOnly: false, isStatic: closureKind == ClosureKind.Singleton);
-                        CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(translatedLambdaContainer, cacheField);
-                        cache = F.Field(receiver, cacheField.AsMember(constructedFrame)); //NOTE: the field was added to the unconstructed frame type.
+                        // We cannot reference a type argument from the parent method outside the method
+                        // so if the lambda has such an argument we cannot cache it in a field
+                        if (!HasTypeArgumentsFromReferencedMethod(cacheVariableType))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1603]: >!HasTypeArgumentsFromReferencedMethod(cacheVariableType) [](start = 28, length = 56)

It looks like we should still be able to cache in a loop. Could you please add a test for a scenario like that?

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 2]
```diff
@@ -1630,6 +1636,39 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
             }
 
             return result;
+
+            bool HasTypeArgumentsFromReferencedMethod(TypeSymbol symbolToCheck)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1640]: >HasTypeArgumentsFromReferencedMethod [](start = 17, length = 36)

Type checks for specific symbol types in this function look suspicious. It would be good to add a comment explaining what exactly we are trying to accomplish here.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 1 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1603]: Done.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs [hunk 3]
```diff
@@ -5377,5 +5377,46 @@ .maxstack  3
   IL_0059:  ret
 }");
         }
+
+        [WorkItem(44720, "https://github.com/dotnet/roslyn/issues/44720")]
+        [Fact]
+        public void LambdaInsideGenericLocalFunctionInsideLoop()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs]: >LambdaInsideGenericLocalFunctionInsideLoop [](start = 20, length = 42)

Given this test scenario. Could you please explain what was going wrong? What types and methods are involved and how the change fixes the problem?

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 2 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1640]: Will do that.

Also it would be nice if somebody review the function very closely. It looks too complicated for me but I haven't found an easier way to check the type.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs [hunk 4]
```diff
@@ -5375,6 +5375,104 @@ .maxstack  3
...
+    private static void Main()
+    {
+        TestMethod(string.Empty);
+    }
+
+    private static void TestMethod<T>(T param)
+    {
+        var message = string.Empty;
+
+        for (int i = 0; i < 1; i++)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5400]: >for (int i = 0; i < 1; i++) [](start = 8, length = 27)

Is there an affected scenario that doesn't involve a loop?

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1640]: > Also it would be nice if somebody review the function very closely. It looks too complicated for me but I haven't found an easier way to check the type.

We will be able to do that once we know what this function is supposed to do

---
In reply to: [436855315](https://github.com/dotnet/roslyn/pull/44939#discussion_r436855315) [](ancestors = 436855315)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs (hunk 4 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5400]: According to [the condition (shouldCacheForStaticMethod || shouldCacheInLoop)](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs#L1586) there should be a scenario involving a static method instead of a loop. Haven't tried though.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 2 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1654]: Add a comment describing the method.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs (hunk 4 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5400]: No, it seems to be reproducable only using a loop. Roslyn does not cache the delegate for static methods if the method is generic (wondering if that can be changed with the check I've added).

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs (hunk 3 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs]: Added the test description.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5413]: > No, it seems to be reproducable only using a loop. Roslyn does not cache the delegate for static methods if the method is generic (wondering if that can be changed with the check I've added).

Is it possible that ```shouldCacheForStaticMethod``` is always set correctly and only ```shouldCacheInLoop``` is incorrect and we should adjust the place where it is calculated to do the same thing as we do while calculating ```shouldCacheForStaticMethod```?


---
In reply to: [437690642](https://github.com/dotnet/roslyn/pull/44939#discussion_r437690642) [](ancestors = 437690642)

**Comment by AlekseyTs** [comment]
>             if (shouldCacheForStaticMethod || shouldCacheInLoop)

I would expect us to figure out if and what kind of caching should be performed by this point and the locals (```shouldCacheForStaticMethod```, ```shouldCacheInLoop```) already reflecting that. 

---
Refers to: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1587 in 8801674. [](commit_id = 88016744a000512746947be866dba6c832d65550, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1654]: > Added a comment describing the method.

It looks like you simply translated C# inside the method to English. What is the referenced method? What is the symbol to check? What kind of dependency between them we are trying to detect?

---
In reply to: [437689304](https://github.com/dotnet/roslyn/pull/44939#discussion_r437689304) [](ancestors = 437689304)

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1603]: >HasTypeArgumentsFromReferencedMethod [](start = 29, length = 36)

It feels like the helper you are looking for is ```public static bool ContainsTypeParameter(this TypeSymbol type, MethodSymbol parameterContainer)``` from TypeSymbolExtensions.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1603]: >HasTypeArgumentsFromReferencedMethod(cacheVariableType) [](start = 29, length = 55)

I would prefer this logic to be pulled out of the entire ```if (shouldCacheForStaticMethod || shouldCacheInLoop)``` and be incorporated into calculation of ```shouldCacheInLoop``` value.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 5]
```diff
@@ -1597,17 +1598,23 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
...
-                            // no need to include the top-level method ordinal in the field name.
-                            (closureKind == ClosureKind.General) ? -1 : topLevelMethodId.Ordinal,
-                            topLevelMethodId.Generation,
-                            lambdaId.Ordinal,
-                            lambdaId.Generation);
-
-                        var cacheField = new SynthesizedLambdaCacheFieldSymbol(translatedLambdaContainer, cacheVariableType, cacheVariableName, _topLevelMethod, isReadOnly: false, isStatic: closureKind == ClosureKind.Singleton);
-                        CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(translatedLambdaContainer, cacheField);
-                        cache = F.Field(receiver, cacheField.AsMember(constructedFrame)); //NOTE: the field was added to the unconstructed frame type.
+                        // We cannot reference a type argument from the parent method outside the method
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1601]: >type argument [](start = 49, length = 13)

a "type parameter"?

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 6]
```diff
@@ -1597,17 +1598,23 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
...
+                        // so if the lambda has such an argument we cannot cache it in a field
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1602]: >if the lambda has such an argument [](start = 30, length = 34)

"if the delegate type refers to such type parameter"?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs [hunk 7]
```diff
@@ -5375,6 +5375,117 @@ .maxstack  3
...
+        {
+            void LocalMethod<TLocal>(TLocal value)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5415]: >void LocalMethod<TLocal>(TLocal value) [](start = 12, length = 38)

I think we should test more permutations and whether caching is used for them:
- both the lambda and the loop are inside a local function 
- This local function is not generic and is inside another local function, the used type parameter belongs to the outer local function rather than to this one,   
- there is no local function, the delegate refers to the type parameter of the enclosing method
- the type parameter of the local function is not used in the delegate type
- etc.

Obviously, if you plan to stick with new ```HasTypeArgumentsFromReferencedMethod``` method, we would want to test more delegate types to cover full breadth of different type symbol kinds.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

**Comment by agocke** [comment]
I'm not sure this is the right fix -- we could try to avoid capturing the cache field in the case that we're capturing a type parameter from an outside scope (which would appear in the type of the cache field and thus be invalid), or we could track captured types as well as captured variables, and if we find one, we can introduce a closure environment that is parameterized on the captured type, which would allow us to successfully create a caching field of the captured type.

**Comment by AlekseyTs** [comment]
@agocke 

> I'm not sure this is the right fix.

Are you saying the fix is incorrect, i.e. is going to cause incorrect behavior or a crash at runtime? Otherwise it looks like a viable alternative to the current behavior - an invalid code is emitted, the program cannot be executed. Perhaps the code will be not as optimal as it could potentially be, but at least it will not crash. We could file a follow up issue to optimize the rewrite later.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 1 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs]: Wow, awesome! That's the method I was looking for. Thanks for the hint.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 5 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 2 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs]: Not actual anymore since I've deleted the method.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs (hunk 4 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5413]: I've dug into the logic behind the second part of the condition (caching into a local instead of a field), and it seams to be a dead code.

`F.SynthesizedLocal` is always invoked with `syntax = null` and `kind = CachedAnonymousMethodDelegate` ([code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs#L1615)) and that leads to failed debug assertion in `SynthesizedLocal..ctor` ([code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedLocal.cs#L47)) or `NullReferenceException` in `LambdaUtilities.GetDeclaratorPosition` ([code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs#L479)).

Considering that, the field is the only possible option to do caching, I can remove the dead code and move my check outside the if statement.

Would appreciate if you confirm if that code is a dead code.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5413]: > Would appreciate if you confirm if that code is a dead code.

At the moment, we are really heads down trying to complete the remaining feature work. One way to confirm that your assumption is incorrect is to make a change to throw when the dead code is reached, run all tests and see if anything breaks. If nothing breaks, there is still a chance that there is a test gap.

Comment (nbalakin) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5413]: Have already done that. None of the tests executes that branch. Haven't tried dogfooding the compiler into itself yet though.

I've tried to come up with a test case to cover that code. That's how I concluded that the code just cannot be executed successfully at all. 🙂

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 8]
```diff
@@ -1597,17 +1598,33 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
...
-
-                        var cacheField = new SynthesizedLambdaCacheFieldSymbol(translatedLambdaContainer, cacheVariableType, cacheVariableName, _topLevelMethod, isReadOnly: false, isStatic: closureKind == ClosureKind.Singleton);
-                        CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(translatedLambdaContainer, cacheField);
-                        cache = F.Field(receiver, cacheField.AsMember(constructedFrame)); //NOTE: the field was added to the unconstructed frame type.
+                        // Also we want to substitute the type according to the type map for the method that uses the delegate.
+                        // Such substitution handles cases when the delegate type uses a type parameter from the method
+                        // and the parameter is not captured by the container.
+                        if (TryGetGenericMethodTypeMap(referencedMethod.ConstructedFrom, out var typeMap))
+                        {
+                            cacheVariableType = typeMap.SubstituteType(cacheVariableType).Type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1606]: >cacheVariableType = typeMap.SubstituteType(cacheVariableType).Type; [](start = 28, length = 67)

Why is this needed? it doesn't look like we did this before.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 9]
```diff
@@ -1597,17 +1598,33 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
                         // different from the local variable `type`, which has the node's type substituted for the current container.
                         var cacheVariableType = containerAsFrame.TypeMap.SubstituteType(node.Type).Type;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1599]: >containerAsFrame.TypeMap.SubstituteType(node.Type).Type; [](start = 48, length = 56)

It looks like this duplicates the work that has already been done. It looks like this what we store in the ```type``` local.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 8 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1606]: After closer investigation, I am convinced that we are approaching this in a somewhat wrong way. After debugging the test, I realized that ```referencedMethod` is the lambda that is being converted to the delegate, it doesn't really provide a context for the delegate type, the context comes from an enclosing method and enclosing types. In terms of the comment below: "We cannot reference type parameters from the parent method outside the method", the ```referencedMethod``` is not a parent method.
Even though the check below appears to provide expected value in our scenarios, this is not the question we should be asking. The better (and simpler) question to ask is whether the ```cacheVariableType``` as calculated above refers to any method type parameters (I mean from any method) after the substitution. If that is the case, we cannot use it for a field in any type. I don't think we have a ready to use helper to perform a check like that, but it should be pretty straight forward to add one next to ```ContainsTypeParameter``` used below following the same implementation strategy.

---
In reply to: [447782927](https://github.com/dotnet/roslyn/pull/44939#discussion_r447782927) [](ancestors = 447782927)

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs [hunk 10]
```diff
@@ -1630,6 +1646,30 @@ private BoundNode RewriteLambdaConversion(BoundLambda node)
             }
 
             return result;
+
+            static bool TryGetGenericMethodTypeMap(MethodSymbol method, out TypeMap typeMap)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1650]: >TryGetGenericMethodTypeMap [](start = 24, length = 26)

Based on my suggestion above, I think this method is not needed.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 6)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenClosureLambdaTests.cs:5413]: > Have already done that. None of the tests executes that branch. Haven't tried dogfooding the compiler into itself yet though.

> I've tried to come up with a test case to cover that code. That's how I concluded that the code just cannot be executed successfully at all.

I suggest opening an issue about that and capturing results of your investigation there. I don't think we should do anything about that in context of this PR because it doesn't appear to be related to the issue we are trying to address.

---
In reply to: [445836446](https://github.com/dotnet/roslyn/pull/44939#discussion_r445836446) [](ancestors = 445836446)

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs]: > I would prefer this logic to be pulled out of the entire if (shouldCacheForStaticMethod || shouldCacheInLoop) and be incorporated into calculation of shouldCacheInLoop value.

I changed my mind, let's not focus on this and keep the fix simple and local.

---
In reply to: [439131915](https://github.com/dotnet/roslyn/pull/44939#discussion_r439131915) [](ancestors = 439131915)

**Comment by RikkiGibson** [comment]
Hi @nbalakin, is anything we can do to help with this PR? We got a duplicate report of the same bug referenced in this PR's description, so it seems people are still hitting the issue from time to time.

**Comment by nbalakin** [comment]
Hi @RikkiGibson. Yep, I can finish the fix.
Could you confirm if I understand your review processes well. All the review remarks that has "#Closed" in the first message are considered to be resolve by the remark author, aren't they? So the only unfixed remarks are those three by AlekseyTs:
- https://github.com/dotnet/roslyn/pull/44939#discussion_r447973647
- https://github.com/dotnet/roslyn/pull/44939#discussion_r439141496
- https://github.com/dotnet/roslyn/pull/44939#discussion_r447974246

 Can I mark them as the other as resolved in GitHub UI just to make the PR  cleaner?

**Comment by 333fred** [comment]
> Can I mark them as the other as resolved in GitHub UI just to make the PR cleaner?

Yes, that's fine. CodeFlow, while a cleaner review UI than github's native tools, does leave things like that to be desired.

**Comment by AlekseyTs** [comment]
> Could you confirm if I understand your review processes well. All the review remarks that has "#Closed" in the first message are considered to be resolve by the remark author, aren't they? 

> Can I mark all the others as resolved in GitHub UI just to make the PR cleaner?

Usually we prefer that only those who made a comment close it (this simplifies the book keeping and makes it easier to make sure that the concerns are indeed addressed). However, I believe there is a way to mark a comment as Resolved (rather than Closed) by adding "#Resolved" in the same place.

File: src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs (hunk 8 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1642]: > The better (and simpler) question to ask is whether the cacheVariableType as calculated above refers to any method type parameters (I mean from any method) after the substitution. 

Consider the following code. In this case the lambda refers to another method type parameter but there is nothing wrong to cache the delegate because the type parameter has larger scope than the cache.
```
void GenericMethod<T>(T value)
{
    for (int i = 0; i < 1; i++)
    {
        DoStuff<T>(() => value);
    }
}

void DoStuff<T>(Func<T> lambda)
{
    Console.WriteLine(lambda());
}
```

I'm totally ok to implement the fix in any way you think is the best. But I'm worried with this kind of fix, because the fix you offer eliminates the cache in that scenario hence may slow down some code that already works fine.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs:1642]: > Consider the following code. In this case, the lambda refers to another method type parameter, but there is nothing wrong to cache the delegate because the type parameter has a larger scope than the cache.

The concept of a "larger scope" for a type parameter probably needs a clear definition. And I think the suggested fix can esily be adjusted to ignore type parameters like that. However, if we are talking about method type parameters that are becoming type type parameters, then, perhaps, the "after the substitution" is the important part in the suggestion. Is it possible that "after the substitution" they become type type parameters and, therefore, are no longer treated as "bad" by the suggested fix? It has been a long time since I looked at this PR, so it is quite possible I am missing something.

Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Lowering/ClosureConversion/ClosureConversion.cs]: I've implemented the fix you offered. Please, take a look.

What about the case I mentioned, you fix handles it perfectly fine. The lambda is already "moved" into a display class, so from the perspective of the fix the type argument of `GenericMethod` is actually a type argument of the display class.
I've added a unit test to verify that such a case works fine after the fix.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 11]
```diff
@@ -1031,6 +1031,15 @@ public static bool ContainsTypeParameters(this TypeSymbol type, HashSet<TypePara
         private static readonly Func<TypeSymbol, HashSet<TypeParameterSymbol>, bool, bool> s_containsTypeParametersPredicate =
             (type, parameters, unused) => type.TypeKind == TypeKind.TypeParameter && parameters.Contains((TypeParameterSymbol)type);
 
+        public static bool ContainsTypeParameterFromSymbol(this TypeSymbol type, Func<Symbol, bool> containingSymbolFilter)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1034]: >ContainsTypeParameterFromSymbol [](http://example.com/codeflow?start=27&length=31)

It looks like there is only one usage of this helper. Generalization of this helper with a filter delegate doesn't feel necessary. I think a specialized helper targeting specific check we need will be prefered, something called "ContainsMethodTypeParameter" and the condition could be as simple as ```(type as TypeParameterSymbol)?.TypeParameterKind == TypeParameterKind.Method```

**Comment by AlekseyTs** [comment]
Done with review pass (commit 11). Overall, the change looks good to me. However, we are still debating whether we would like to merge it and open a separate issue to enable proper caching for the affected scenarios, or would like to go with an alternative fix that fixes the caching rather than disables it (we don't have a PR for that yet).

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 11 above)
Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: I've made the method more specialised but I've left my implementation of the check (`ContainingSymbol is MethodSymbol` instead of `TypeParameterKind == Method`)

Not sure about correctness of my check if my check in terms of correct usage of code model, but the current implementation handles the case we discussed earlier. Consider the following:

Original code:
```
void GenericMethod<T>(T value)
{
    for (int i = 0; i < 1; i++)
    {
        DoStuff<T>(() => value);
    }
}

void DoStuff<T>(Func<T> lambda)
{
}
```
Generated code:
```
private void GenericMethod<T>(T value)
{
    DisplayClass<T> displayClass = new DisplayClass<T>();
    displayClass.value = value;
    for (int i = 0; i < 1; i++)
    {
        DoStuff(new Func(displayClass.Lambda));
    }
}
```
When the compiler chooses if to cache the delegate, the type parameter symbol has `TypeParameterKind == Method` (as it is in the original parameter symbol) but the container is an instance of `SynthesizedClosureEnvironment` as the parameter is moved to the synthesised display class.

Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: BTW, should I re-request a review via GitHub button when I think I fixed all the remarks? Or do you monitor all the updates made somehow else?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: >  should I re-request a review via GitHub button when I think I fixed all the remarks? Or do you monitor all the updates made somehow else?

There is no requirement to do this, however you may re-request a review when you feel appropriate. We don't use any special system to track active PRs other than what GitHub provides. So, if there are many active PRs, a PR might fall of the radar. When this happens, it is recommended to bring our attention to the PR.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: > I've left my implementation of the check (ContainingSymbol is MethodSymbol instead of TypeParameterKind == Method)

That is fine, both forms should be equivalent.

> When the compiler chooses if to cache the delegate, the type parameter symbol has TypeParameterKind == Method (as it is in the original parameter symbol) but the container is an instance of SynthesizedClosureEnvironment as the parameter is moved to the synthesised display class.

It sounds like you are saying that you have run into a case when a type type parameter has ```TypeParameterKind == TypeParameterKind.Method```. If that is the case, this is a bug.

Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: > It sounds like you are saying that you have run into a case when a type type parameter has TypeParameterKind == TypeParameterKind.Method. If that is the case, this is a bug.

Yep, that is the case I have run into.
It happens because the original type parameter is wrapped by `SynthesizedSubstitutedTypeParameterSymbol` ([code](https://github.com/dotnet/roslyn/blob/1cca63b/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSubstitutedTypeParameterSymbol.cs#L18)) that passes some properties through the original type parameter symbol (e.g. `TypeParameterKind`), but some are replaced with a new implementation (e.g. `ContainingSymbol`).

I cannot judge if it is the correct behaviour, but if it is not the correct one we can create and issue and I can fix it. And to fix it I will need some help to decide what is the expected behaviour in such cases.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: > I cannot judge if it is the correct behaviour, but if it is not the correct one we can create an issue and I can fix it. But to fix it I will need some help to decide what is the expected behaviour in such cases.

I created a PR [https://github.com/dotnet/roslyn/pull/53968](https://github.com/dotnet/roslyn/pull/53968) with asserts that represent expected behavior. The asserts are failing for some scenarios. I think the fix for SynthesizedSubstitutedTypeParameterSymbol is to override TypeParameterKind property as it is done in SynthesizedClonedTypeParameterSymbol for VB.

Comment (nbalakin) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: I'll use your branch and implement the fix.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 13)

**Comment by AlekseyTs** [comment]
> we are still debating whether we would like to merge it and open a separate issue to enable proper caching for the affected scenarios, or would like to go with an alternative fix that fixes the caching rather than disables it

Given that the fix is very targeted and affects only scenarios that would produce invalid metadata, we decided to merge it.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Need a second sign off on a community PR.

**Review by RikkiGibson — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@nbalakin Thanks for the contribution.
